shards/sched.go

```diff
@@ -0,0 +1,230 @@
+// search is a RLock.
+//
+// ### Co-operatively limit long running searches
+//
+// Some searches are slow. Either due to a hard to execute search query (can't
+// use trigram index) or a large number of results. We want to support this
+// use case while still allowing interactive queries to be fast.
+//
+// ### No tuneables
+//
+// We want to avoid the need to tune the scheduler depending on the workload /
+// instance. As such we use a simple design whose inputs are time and number
+// of CPUs.
+//
+// ## Design
+//
+// We use semaphores to limit the number of running processes. An exclusive
+// process acquires the full semaphore. A process represents something which
+// has acquired on the semaphore. Every process is either fast or slow. A
+// process starts as fast, but is downgraded to slow after a period of
+// time. Downgrading relies on a process co-operatively deciding to downgrade.
+//
+// We intentionally keep the algorithm simple, but have a general interface to
+// allow improvements as we learn more.
+type scheduler struct {
+	// Exclusive process holds a write lock, search processes hold read locks.
+	mu             sync.RWMutex
+	semInteractive *semaphore.Weighted
+	semBatch       *semaphore.Weighted
+
+	// interactiveDuration is how long we run a search query at interactive
+	// priority before downgrading it to a batch/slow query.
+	interactiveDuration time.Duration
+}
+
+func newScheduler(capacity int64) *scheduler {
+	// Burst upto 1/4 of CPUs for batch. This means we now can use upto 1.25 the
+	// amount of CPU.
+	batchCap := capacity / 4
+	if batchCap == 0 {
+		batchCap = 1
+	}
+
+	return &scheduler{
+		semInteractive: semaphore.NewWeighted(capacity),
+		semBatch:       semaphore.NewWeighted(batchCap),
+
+		interactiveDuration: 5 * time.Second,
+	}
+}
+
+// Acquire blocks until a normal process is created (ie for a search
+// request). See process documentation. It will only return an error if the
+// context expires.
+func (s *scheduler) Acquire(ctx context.Context) (*process, error) {
+	s.mu.RLock()
+
+	// Start in interactive. yieldFunc will switch us to batch. sem can be nil
+	// if we fail while switching to batch. nil value prevents us releasing
+	// twice.
+	sem := s.semInteractive
+
+	if err := sem.Acquire(ctx, 1); err != nil {
+		s.mu.RUnlock()
+		return nil, err
+	}
+
+	return &process{
+		releaseFunc: func() {
+			if sem != nil {
+				sem.Release(1)
+				sem = nil
+			}
+			s.mu.RUnlock()
+		},
+		yieldTimer: newDeadlineTimer(time.Now().Add(s.interactiveDuration)),
+		yieldFunc: func(ctx context.Context) error {
+			if sem != nil {
+				sem.Release(1)
+				sem = nil
+			}
+
+			// Try to acquire batch. Only set sem if we succeed so we know we can
+			// clean it up. If this fails we assume the process will stop running
+			// (ctx has expired).
+			semNext := s.semBatch
+			if err := semNext.Acquire(ctx, 1); err != nil {
+				return err
+			}
+
+			sem = semNext
```

Comment:
We don't have a good way of seeing whether this actually works in production. Should we add a little log here so that we can see if processes get "downgraded"?

Reply:
Maybe this is also a good spot to log the fill levels of the two semaphores?

Reply:
I want to add observability before landing this PR. This will be part of it. I suppose this is deserving of a log line since it should happen much less than an interactive search.